### PR TITLE
Lower std::string's alignment requirement from 16 to 8.

### PR DESCRIPTION
**Pull request by EricWF**
This allows smaller allocations to occur, closer to the actual std::string's required size. This is particularly effective in decreasing the allocation size upon initial construction (where __recommend is called to determine the size).

Although the memory savings per-string are never more than 8 bytes per string initially, this quickly adds up. And has lead to not insigficant memory savings at Google.

Unfortunately, this change is ABI breaking because it changes the value returned by max_size. So it has to be guarded.

### CODE REVIEW

**Comment by EricWF** [comment]
This is a duplicate of #68749 which was closed by accident.

File: libcxx/test/libcxx/strings/basic.string/string.capacity/allocation_size.pass.cpp
Comment (mordante) [inline libcxx/test/libcxx/strings/basic.string/string.capacity/allocation_size.pass.cpp:20]: This comment seems wrong, the value 16 is no longer hard-coded.

**Review by mordante — COMMENTED** [review]
In general no objection.
We should add this to the release notes too.

File: libcxx/test/libcxx/strings/basic.string/string.capacity/allocation_size.pass.cpp
Comment (EricWF) [reply libcxx/test/libcxx/strings/basic.string/string.capacity/allocation_size.pass.cpp]: Done.

File: libcxx/test/libcxx/strings/basic.string/string.capacity/max_size.pass.cpp
Comment (EricWF) [reply libcxx/test/libcxx/strings/basic.string/string.capacity/max_size.pass.cpp:22]: I would rather revisit this after landing the change.
There's more risk and more reason to reject the changes if we apply it globally.
So let's get this in, then consider if we can push it to everybody.

Comment (mordante) [reply libcxx/test/libcxx/strings/basic.string/string.capacity/max_size.pass.cpp:22]: I'm fine with that too, as mentioned I was mainly curious.

Comment (ldionne) [reply libcxx/test/libcxx/strings/basic.string/string.capacity/max_size.pass.cpp:22]: Can you prepare the patch to clean up the ABI flag and make it the default value before we merge this (kind of like a stacked review even though those aren't great in GH PRs)? I'd really like to make sure that we don't just check in this ABI flag without its removal if we all agree that we don't want the ABI flag in the end.

**Comment by EricWF** [comment]
OK, I have no idea how to rebase diffs apparently. Working on fixing this.

File: libcxx/test/libcxx/strings/basic.string/string.capacity/max_size.pass.cpp
Comment (EricWF) [reply libcxx/test/libcxx/strings/basic.string/string.capacity/max_size.pass.cpp:23]: It was easier for me to write that patch on top  of HEAD, rather than this change. But here it is as requested.
#68925

**Comment by EricWF** [comment]
> In general no objection. We should add this to the release notes too.

Sorry, I wasn't ignoring this. But the release note should be there now.

File: libcxx/include/__config
Comment (ldionne) [inline libcxx/include/__config:170]:
```suggestion
// Save memory by providing the allocator more freedom to allocate the most
```

File: libcxx/docs/ReleaseNotes/18.rst
Comment (ldionne) [inline libcxx/docs/ReleaseNotes/18.rst:140]:
```suggestion
  to throw a different exception when attempting allocations that are too large
```

File: libcxx/test/libcxx/strings/basic.string/string.capacity/allocation_size.pass.cpp
Comment (ldionne) [inline libcxx/test/libcxx/strings/basic.string/string.capacity/allocation_size.pass.cpp:42]: Can you explicitly `return 0` from the test? Otherwise this fails under `-ffreestanding` because `main` isn't a special function then.

**Review by ldionne — APPROVED** [review]
Thanks, this LGTM with a few nitpicks. The C++03 build is failing but I think this is easy to address (probably your usage of `constexpr` in the test).

File: libcxx/test/libcxx/strings/basic.string/string.capacity/allocation_size.pass.cpp
Comment (EricWF) [reply libcxx/test/libcxx/strings/basic.string/string.capacity/allocation_size.pass.cpp:42]: Done.

File: libcxx/docs/ReleaseNotes/18.rst
Comment (EricWF) [reply libcxx/docs/ReleaseNotes/18.rst:140]: Done.

File: libcxx/include/__config
Comment (EricWF) [reply libcxx/include/__config:170]: lol sorry this snuck back in after I broke broke the last PR.

File: libcxx/docs/ReleaseNotes/18.rst
Comment (mordante) [inline libcxx/docs/ReleaseNotes/18.rst:137]: s/std::string/``std::string``/ so it will be rendered properly, the same for other places.

File: libcxx/docs/ReleaseNotes/18.rst
Comment (mordante) [inline libcxx/docs/ReleaseNotes/18.rst:138]: s/std::string::max_length/``std::string::max_size``

**Comment by mordante** [comment]
> > In general no objection. We should add this to the release notes too.
> 
> Sorry, I wasn't ignoring this. But the release note should be there now.

No problem.

**Comment by mordante** [comment]
sorry wrong button, still getting used to GitHub reviews

**Review by mordante — APPROVED** [review]
LGTM modulo some nits.

**Pull request merged**
